Video: Redux state changes for custom video poster

donnapep opened the pull request:
This PR separates out the Redux state changes from PR #11389, which adds support for selecting or uploading a custom image to serve as the poster for a video.

Post discussing Calypso-based VideoPress poster chooser can be found at pxWta-JO-p2 (internal ref).

[Upload poster endpoint](https://developer.wordpress.com/docs/api/1.1/post/videos/%24guid/poster/)

donnapep commented:
@marcinbot @lamosty @gwwar Tagging you guys to review this one as it looks like you worked on the image editor. This is the video equivalent of that. Thx!

gwwar commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -2350,6 +2350,34 @@ Undocumented.prototype.transferStatus = function( siteId, transferId ) {
+ * @param {Function} fn Function to invoke when request is complete
+ * @returns {Promise} A promise that resolves when the request is complete
+ */
+Undocumented.prototype.updateVideoPoster = function( videoId, data, fn ) {
```
Is this a new wpcom only endpoint? Did we have any plans for Jetpack support?

dmsnell replied:
I'll very much encourage using the new data layer for any and all new wpcom endpoints and would be happy to walk you through it. In the meantime, check out the [API layer](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/README.md) and [HTTP layer](https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom-http/README.md) readmes.

dbtlr replied:
@gwwar This is a WPCOM only endpoint, as VideoPress is a WPCOM feature. Jetpack uploads its videos to WPCOM for all VideoPress requests, not the other way around.

donnapep replied:
@gwwar Asked about this and it's for wpcom only, although it supports Jetpack in the backend. But VideoPress is technically a wpcom feature that Jetpack accesses.

gwwar commented on client/state/ui/editor/video-editor/actions.js:
```diff
@@ -0,0 +1,86 @@
+ * @param  {Object}   params  Poster data
+ * @return {Function}         Action thunk
+ */
+export const updateVideoEditorPoster = ( guid, params ) => {
```
I think it'd be useful to describe the `params` shape in the jsdoc

gwwar replied:
Also, I think returning a thunk here is fine, though if you're up for it, there's a new [data-layer](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer) pattern that we can use. 

The upside is that testing becomes a lot easier, since we can usually dispatch a simple action to trigger the entire the request. Maybe something like
```jsx
dispatch( { type: VIDEO_EDITOR_POSTER_UPDATE, videoId, params } );
```

Ping me or @dmsnell for help on that

gwwar commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,149 @@
+	hasScriptLoadError,
+	poster,
+	posterIsUpdating,
+	posterIsUpdated,
```
What's the difference here between `posterIsUpdating` and `posterIsUpdated`?

donnapep replied:
@gwwar When the poster is being updated on the server, `posterIsUpdating` gets set and the UI will show a spinner.

I had to introduce a separate `posterIsUpdated` because I needed to know if the poster had been successfully updated, and if so, then close out the video editor and return to the previous screen - https://github.com/Automattic/wp-calypso/pull/11648/files#diff-35b626cb9b7af72412a7ab302d8fc3f6R67

dbtlr commented on client/lib/wpcom-undocumented/lib/undocumented.js:
```diff
@@ -2350,6 +2350,34 @@ Undocumented.prototype.transferStatus = function( siteId, transferId ) {
+		];
+	}
+
+	if ( 'at_time' in data ) {
```
I would make this with an `elseif`, since you can't send both (the logic will always take the `at_time` first on the server side). Also, you should have an `else` at the end, that if neither is present, it will cause this to not post to the API.

dmsnell replied:
@dbtlr the way we would normally prefer to handle this isn't with a dangling `else` at the end but rather with an early abort from the function call. you will find that we have very few `else` structures in Calypso.

also, in general, I would strongly encourage working together on this to use the new data layer instead of creating a new `wpcom.undocumented` function

donnapep replied:
@dmsnell Re: using the data layer. It's currently in progress. ;)

dmsnell replied:
@donnapep please don't hesitate to discuss the work! the data layer is still very new and we're still iterating on the developer experience, so I would value your input

dbtlr commented on client/state/ui/editor/test/reducer.js:
```diff
@@ -16,7 +16,8 @@ describe( 'reducer', () => {
 			'contactForm',
-			'imageEditor'
+			'imageEditor',
+			'videoEditor',
```
Is it proper to have a final comma in a JavaScript array?

dmsnell replied:
in Calypso it is yeah. we encourage it for the effects on diffs

dbtlr commented on client/state/ui/editor/actions.js:
```diff
@@ -13,7 +13,8 @@ export const MODAL_VIEW_STATS = {
 	[ ModalViews.GALLERY ]: 'view_gallery',
-	[ ModalViews.IMAGE_EDITOR ]: 'view_edit'
+	[ ModalViews.IMAGE_EDITOR ]: 'view_edit',
+	[ ModalViews.VIDEO_EDITOR ]: 'view_edit',
```
comma?

dbtlr commented on client/state/ui/editor/video-editor/actions.js:
```diff
@@ -0,0 +1,88 @@
+	VIDEO_EDITOR_SCRIPT_LOAD_ERROR,
+	VIDEO_EDITOR_STATE_RESET,
+	VIDEO_EDITOR_STATE_RESET_POSTER,
+	VIDEO_EDITOR_VIDEO_HAS_LOADED,
```
comma?

donnapep replied:
@dbtlr I like to do this for objects so that if you need to add another property later, you don't get the extra diff on the preceding line. I see this being done occasionally in Calypso, but it's not part of the style guide.

donnapep commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,51 @@
+		method: 'POST',
+		path: `/videos/${ action.videoId }/poster`,
+		body: 'at_time' in action.params ? action.params : undefined,
+		formData: 'file' in action.params ? [ [ 'poster', action.params.file ] ] : undefined,
```
@dmsnell `formData` does not appear to be sent with the POST request. Before I get into some deeper debugging, does what I have here so far look OK?

dmsnell replied:
possibly a bug? I will look into this

donnapep replied:
A fix for this can be found in PR #12135, although it still needs more testing to ensure nothing breaks. Thx!

donnapep commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,51 @@
+		onFailure: action,
+	} ) );
+
+	return next( action );
```
@dmsnell I don't quite get when I should use `next` vs. `dispatch`. And in what sort of circumstances would I want to do neither of those things?

dmsnell replied:
`next()` is going to be used when we're passing the given action along the middleware chain. maybe we intercept post editor updates and want to make sure that it's `capital_P_dangit`. we could intercept that action, filter `post_content`, and then we would want to pass it along to the next link in the chain with `next()` - this is like `add_filter()` in WP.

on the other hand, we might prefer to trigger some side-effect or corresponding action. maybe every time we see `Wordpress` we want to bump some state. in this case we would pass along the original untouched action to `next()` but since we're firing off a new action altogether we want it to start at the front of the middleware stack so we'll fire off `dispatch( bumpStat( … ) )` - akin to `add_action` in WP.

@coderkevin just since we recently discussed actions and filters I thought I'd poke you on this ^^^

gwwar replied:
@dmsnell is there ever a case where we wouldn't want to pass the action along the middleware chain? As is we'll get weird oddities like the dispatch action not being recorded by redux dev tools if we omit `return next( action )` in addition to missing any other side effect middleware that was lower on the chain.

For example in console if you dispatch:
```
dispatch( { type: 'ACCOUNT_RECOVERY_RESET_REQUEST' } );
```
We'll see the other actions recorded but not the initial action request :
![screen shot 2017-03-10 at 1 51 54 pm](https://cloud.githubusercontent.com/assets/1270189/23814588/efea2b64-0598-11e7-8cb6-a5f3d9fc36be.png)
https://github.com/Automattic/wp-calypso/pull/11699/files#diff-2c98cce53eb8b5257f1bbca88632820cR17
cc @southp

dmsnell replied:
@gwwar the actions should be recorded on their first `dispatch()` so they should appear even if we don't pass them along (would be a bug if this isn't the case and we would probably just need to rearrange the middleware)

times we wouldn't want to pass along the action? maybe so. we might want to cull actions from hitting the reducers or system for some reason. not sure off-hand what these might be but it seams very plausible to me.

coderkevin replied:
I think this is all well and good, but it's a pretty privileged operation here. I feel more thought should go into what should be allowed for which Calypso extensions when we start accepting more outside contributions.

donnapep commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,51 @@
+};
+
+export default {
+	[ VIDEO_EDITOR_POSTER_UPDATE ]: [ dispatchRequest( updateVideoEditorPoster, updateVideoPosterSuccess, updateVideoPosterError ) ],
```
@dmsnell It seems that I needed to add a new action to be used explicitly by the data layer. Is this true in general?

dmsnell replied:
@donnapep I'm not sure what your question is. the data layer latches into existing actions which feed through the system. it handle side-effects and data fetching.

the idea is that Calypso can function as best as possible even if no data gets synchronized. so maybe we fire an event `VIDEO_POSTER_UPDATE` which specifies a video and a frame for the post (is that what this is about?) then Calypso can hold onto that state the same way it would if it had just loaded that data from the API.

in this case the data layer sees that action fly by and it says to itself, "self, I know that when someone changes the video poster we also need to let WordPress.com know so that it persists and propagates, I will do just that"

now, once we have fully uploaded the data and get some response we can make a judgement. if it's as simple as saying, "use frame 1337 for the poster" then if we get a positive response back from the server showing that it stored 1337 and our state already shows 1337 (because Redux state operates independently of the middleware) then we don't actually need to do a thing. no changes are required. on the other hand, if it comes back with the wrong frame number we may need to respond in one of many possible ways: retry the operation (Calypso meanwhile is unaware of the failure), report the failure to the user, or just force Calypso to update (overwriting the customer's choice).

please let me know if this is addressing your question!

dmsnell commented:
thanks for your rework on this @donnapep!

instead of diving into the details at this moment I'd like to raise some higher-level questions:

 - what is the purpose of storing data points like `isLoading` and `isUpdating`? what visual needs do we have for their representation? is it possible that we can already represent those values _without_ storing additional state?

 - what is the purpose of having `_SUCCESS` and `_FAILURE` actions? I've been pushing hard to keep those out of the data layer and instead write code that focusses on the actual actions being done in the system: `ADD_POST` instead of `POST_RECEIVE_SUCCESS` for example. Although I think it's harder at the time we design the code to think about doing without these `_SUCCESS/FAILURE` actions I also believe that it leads to stronger code (because we decouple the state tree from the API) and smaller sprawl (because we reuse existing code).

 - there's an additional function we can add to `dispatchRequest()` which will be called on progress events. if we need to represent something like percentage of completed upload we can do that specifically with the data layer in an elegant way

 - we may have found a bug if `formData` isn't coming through. I'll try to look into this

doing well - I'm excited to have what I think is our first data uploader in the data layer!

donnapep commented:
@dmsnell

> what is the purpose of storing data points like isLoading and isUpdating? what visual needs do we have for their representation?

`isLoading` is used to check whether or not the video has fully loaded. If it hasn't, the _Select Frame_ button will be disabled since it doesn't make sense to select a frame of a video that hasn't loaded yet. `isUpdating` shows a spinner (or perhaps this will now change to a progress bar) while the poster is being updated.

> what is the purpose of having _SUCCESS and _FAILURE actions?

We'd like to surface an error message if the poster failed to update. But if the update succeeds, then the poster URL is returned by the endpoint and the thumbnail for the video is updated.
 
> there's an additional function we can add to dispatchRequest() which will be called on progress events. if we need to represent something like percentage of completed upload we can do that specifically with the data layer in an elegant way

I saw that. This may work when uploading a custom image, although likely not when selecting a particular frame. I'll play around with it some.

> we may have found a bug if formData isn't coming through. I'll try to look into this

Sounds good. Cheers!

dmsnell commented:
> isLoading is used to check whether or not the video has fully loaded

don't we already get an event dispatched at the point the video has fully loaded? are we showing the customer some indicator that it's loading? this is an upload, right? is there a reason to prefer a binary `isLoading/isNotLoading` over a progress indicator with percentage of upload?

> isUpdating shows a spinner

is this consistent with other Calypso idioms? I thought we normally assume that an operation works and then only undo if it fails. is there something fundamentally different about the poster calling for the behavior? is setting the poster frame a long operation that warrants showing some process state?

> We'd like to surface an error message if the poster failed to update.

Here is where the data layer is designed to support Calypso design. If our goal is to surface an error message, then why not simply surface an error message? We have an action already for that - `createNotice()`…

> if the update succeeds, then the poster URL is returned by the endpoint and the thumbnail for the video is updated.

…continuing on that thought, if our intention is to store the URL for a poster frame, why not use an action that mirrors that intention - `MEDIA_ASSIGN_POSTER_FRAME` or something like that. If we design our actions to be independent of the specifics of the API then we can communicate much more clearly in the code. It will also make it much easier to test with from within the console.

donnapep commented:
> don't we already get an event dispatched at the point the video has fully loaded?

Nope, not for VideoPress videos. This will be added by a separate PR in this commit https://github.com/Automattic/wp-calypso/pull/11653/commits/68a9d5f915d4779ab2ac68f14b92b03bd134882e

> are we showing the customer some indicator that it's loading?

Yes, but not as part of Calypso. VideoPress controls the UI for this.

> this is an upload, right? is there a reason to prefer a binary isLoading/isNotLoading over a progress indicator with percentage of upload?

Nope, not an upload. This is just to initially display the video in the video editor. Once the video is loaded, then we can enable the button that allows the user to select a specific frame to use as the poster (or they can also choose to upload their own image instead of choosing a frame, this is where upload comes into play).

I think I've done a poor job of communicating the context for this PR. To get a better idea of the background, the master PR #11389 might explain it better.

> is this consistent with other Calypso idioms? I thought we normally assume that an operation works and then only undo if it fails.

Honestly, I'm not sure. That's why I rely on reviewers like yourself to school me. 😉 I will say that video editor is largely modelled after image editor. With image editor, if the image can't be updated then an error message is shown onscreen. For consistency, video editor does a similar thing if the poster couldn't be updated. There's nothing to really "undo" though when updating the poster.

> is there something fundamentally different about the poster calling for the behavior? is setting the > poster frame a long operation that warrants showing some process state?

Uploading the image can take awhile, especially for larger images. Selecting a particular frame is generally faster. I do think the upload warrants an indicator (thinking to potentially switch to a progress bar instead of a spinner here), but perhaps nothing is needed for selecting a frame. 

> Here is where the data layer is designed to support Calypso design. If our goal is to surface an error message, then why not simply surface an error message? We have an action already for that - createNotice()…

Thanks! This is worth investigating. 

> …continuing on that thought, if our intention is to store the URL for a poster frame, why not use an action that mirrors that intention - MEDIA_ASSIGN_POSTER_FRAME or something like that. If we design our actions to be independent of the specifics of the API then we can communicate much more clearly in the code. It will also make it much easier to test with from within the console.

I dunno. Maybe. But what if I wanted to do something else after the poster is updated in addition to assigning the poster frame URL? Now my action no longer tells the whole story and is a bit of a misnomer.

dmsnell commented:
> But what if I wanted to do something else after the poster is updated in addition to assigning the poster frame URL? Now my action no longer tells the whole story and is a bit of a misnomer. I find this naming convention a bit less future-proof.

If we need to do more we can dispatch more actions. The point is to keep our actions specific to the intentions they convey. Ideally, if we have multiples visual changes that need to happen in response to the event then we will find some aptly-named action (such as `SET_POSTER_FRAME`) being listened for in multiple reducers. The predominant model I find is is one big reducer pasting in some `GET_THING_SUCCESS` action and it changes multiple independent state values.

dmsnell commented:
> are we showing the customer some indicator that it's loading?
> > Yes, but not as part of Calypso. VideoPress controls the UI for this.

In this case I would probably argue that we don't need to be storing the `isLoading` state since it sounds like it's already handled externally

donnapep commented:
> In this case I would probably argue that we don't need to be storing the isLoading state since it sounds like it's already handled externally

While true, there is also a need to disable a button until the video is fully loaded, and that _is_ part of Calypso UI.

dmsnell commented:
> difficult to decide when to use React vs. Redux state

interesting. I'm not sure I follow what you mean. do you mean storing state internally to a component via `this.state`? even Redux values come in through `this.props`

> disable a button until the video is fully loaded

that's fair. just wanting to make sure that scrutinize all of the state coming in to the system.

donnapep commented:
> interesting. I'm not sure I follow what you mean. do you mean storing state internally to a component via this.state? even Redux values come in through this.props

Yes, using `this.setState` instead of Redux state. I did end up moving some things like whether or not the video has loaded from Redux to component state. But was it necessary? Does it matter?

I like what Abramov said about this [here](https://github.com/reactjs/redux/issues/1287#issuecomment-175351978). Following that advice, sounds like none of this data necessarily needs to be stored in Redux state. But because we want to hit an endpoint, it was just easier (i.e. "less awkward") to use Redux, for some part of it at least.

dmsnell commented:
> none of this data necessarily needs to be stored in Redux state

my rule of thumb is this: "_*If it should persist across a page refresh store it in Redux*_" also I want to store the _minimum_ amount of data necessary to fully render Calypso to the user. if more than one component may trigger updates to the data that's also an almost immediate indicator that it should be in Redux, or if it's tied to remote data in any way

donnapep commented:
@dmsnell This is ready for another review. I eliminated some of the Redux actions by using component state instead, and renamed the other actions to be more descriptive. I also added a progress action that will be used for showing a progress bar when uploading.

I tried using 'createNotice()` for displaying an error message, but they don't play well with modals so I've retained the action for showing an error.

Thx!

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,67 @@
+ * @returns {Object} original action
+ */
+export const updatePoster = ( { dispatch }, action, next ) => {
+	if ( ! ( 'file' in action.params ) && ! ( 'at_time' in action.params ) ) {
```
just for fun, we can also flip this boolean expression and to some people (me included) it would be more obvious what we're checking. to others it won't be. in general, there will be times where it can dramatically simplify the logic. for more info, check out boolean reduction and _deMorgan's_ law

```js
if ( ! ( 'file' in action.params || 'at_time' in action.params ) ) {
	return next( action );
}
```

further, if we know we would want to use these anyway, we can actually go ahead and extract them in a way which yet again can lead to more insight into our goals

```js
const { at_time, file } = pick( action.params, [ 'at_time', 'file' ] );

if ( ! ( at_time && file ) ) {
	return next( action );
}
```

now these values are already bound for us

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,67 @@
+
+	if ( 'at_time' in action.params ) {
+		params.body = action.params;
+	}
```
if we use `Object.assign()` we can merge these imperative assignments into one declarative list and get rid of those `if` structures

```js
const params = Object.assign(
	{
		apiVersion: '1.1',
		method: 'POST',
		path: …
	},
	file && { formData: [ [ 'poster', file ] ] },
	at_time && { body: action.params },
);
```

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	if ( 'at_time' in action.params ) {
+		params.body = action.params;
```
is this line correct? it seems suspicious. if it _is_ right, we might need a comment to explain _why_ since it looks like it could be a bug

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,67 @@
+export const updateUploadProgress = ( { dispatch }, action, next, progress ) => {
+	let percentage = 0;
+
+	if ( 'loaded' in progress && 'total' in progress ) {
```
these should be guarantees on the HTTP layer, so if an event comes back without them then we have discovered a bug. did you find this to be true?

if the guarantee holds then I don't see any reason why we would need to make sure they're here

donnapep replied:
I saw in [this example](https://developer.mozilla.org/en/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest#Monitoring_progress) that it's possible for `total` to be unknown if `lengthComputable` is `false`, so I figured better safe than sorry and opted to check them both. Although I suppose a better solution may be to do a check on `lengthComputable` in the HTTP layer and eliminate the check in the handler. Thoughts?

dmsnell replied:
yeah we should definitely handle that HTTP-related guarantee in the HTTP layer so that every other middleware in Calypso doesn't have to duplicate the logic. this is all the fun of working with standards :smile:

dmsnell replied:
also @donnapep I think we may have some choices in those cases where `total` might be something like `Infinity` if `lengthComputable` is `false`. that should still indicate as much where progress is concerned and it will give consumers the ability to check if they need to

donnapep replied:
Not sure I follow your last comment. Are you saying that the `lengthComputable` check _shouldn't_ be done in the HTTP layer and is fine as is, because the consumer will at least know that the file is being currently being uploaded?

dmsnell commented on client/state/ui/editor/video-editor/actions.js:
```diff
@@ -0,0 +1,75 @@
+	return {
+		type: VIDEO_EDITOR_RESET_STATE,
+	};
+};
```
fun fact! we can use arrow functions without `{}`s here since all we're doing is returning a value

```js
export const resetState = () => ( {
	type: VIDEO_EDITOR_RESET_STATE,
} );
```

notice that because _objects_ and _function bodies_ share the curlies, I had to surround the returned value in parentheses. this is the only time the arrow-function syntax demands them

dmsnell commented on client/state/ui/editor/video-editor/actions.js:
```diff
@@ -0,0 +1,75 @@
+ *
+ * @param {String} videoId  ID of the video
+ * @param {Object} params  Poster data
+ * @param {Number} [params.at_time]  Number of seconds into the video at which to get the poster
```
since we're now introducing a new action in the Calypso/Redux idiom, why not follow our conventions and use `atTime` instead of `at_time`? The fact that we have to know, "Oh shoot, this function expects a value that looks like something the API wants…" makes me think that we're likely coupling concerns which would be best left separate.

Thanks for documented the properties though!

dmsnell commented on client/state/ui/editor/video-editor/actions.js:
```diff
@@ -0,0 +1,75 @@
+		type: VIDEO_EDITOR_SHOW_UPLOAD_PROGRESS,
+		percentage,
+	};
+};
```
having this as an action in Redux separate from the API is so great because now if we want to test it out we can open up our console and write, `dispatch( { type: VIDEO_EDITOR_SHOW_UPLOAD_PROGRESS, percentage: 57 } )` and suddenly our UI will reflect a 57% progress…

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action object
+ * @return {Object}        Updated state
```
it actually looks like this returns a `Boolean`, and `Updated state` may not be the most descriptive. it's not too difficult since the function name itself documents, but often I like to at least put something like "whether or not a video poster has been updated already" if I put anything at all on the `@returns` tag.

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action object
+ * @return {Object}        Updated state
```
`String`

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action object
+ * @return {Object}        Updated state
```
`Number`

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action object
+ * @return {Object}        Updated state
```
`Boolean`

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,99 @@
+	}
+
+	return state;
+};
```
in this reducer, it looks like we only have a single `true` value - when the action type is `VIDEO_EDITOR_SHOW_ERROR` which could be much more succinctly captured in the following…

```js
export const hasPosterUpdateError = ( state = false, { type } ) => type === VIDEO_EDITOR_SHOW_ERROR;
```

further, what's the point of storing a boolean status value for `hasError`? do we not store _what kind_ of error it is? what value is there in this on/off state?

donnapep replied:
In this case, I don't think we necessarily care what the error is because it's not being used. We only care that the poster update wasn't successful, so that we can in turn communicate that to the user. I suppose we could log the specific error, but other than that it wouldn't be used.

dmsnell commented on client/state/ui/editor/reducer.js:
```diff
@@ -33,6 +34,7 @@ export function postId( state = null, action ) {
 export default combineReducers( {
 	postId,
 	imageEditor,
+	videoEditor,
```
it appears like the video editor state has no reference to any particular media id or site or anything. it's "whatever is out front in the editor" but what happens if we change sites or change videos which we are working on? won't we have race-conditions with network requests on the wire?

donnapep replied:
The video editor is a modal that is opened from the media library modal. Media library modal handles passing the particular video and everything else the video editor needs to know. And because it's a modal, site switching should be a non-issue. Hope that helps clarify and makes sense.

dmsnell commented on client/state/ui/editor/video-editor/selectors.js:
```diff
@@ -0,0 +1,41 @@
+ */
+export function hasPosterUpdateError( state ) {
+	return state.ui.editor.videoEditor.hasPosterUpdateError;
+}
```
please note that we are putting all new selectors into the flat `state/selectors` directory as part of a new search-based selector system (check the devdocs). they belong one for each file where the function name matches the file name except for casing

```js
// has-poster-update-error.js
export default function hasPosterUpdateError( state ) {
	…
}
```

donnapep commented:
@dmsnell The ball's back in your court!

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,88 @@
+	switch ( action.type ) {
+		case VIDEO_EDITOR_SET_POSTER_URL:
+		case VIDEO_EDITOR_SHOW_ERROR:
+			return true;
```
is the poster really updated if we got an error? seems like a semantic mismatch between the data point and its use

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,69 @@
+		return next( action );
+	}
+
+	const { atTime, file } = pick( action.params, [ 'atTime', 'file' ] );
```
I don't think `pick` is necessary here as the destructuring should already do exactly what we want…

```js
const { atTime, file } = action.params;
```

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,69 @@
+};
+
+export const updatePosterError = ( { dispatch } ) => {
+	dispatch( showError() );
```
this error concept is still bugging me. we don't actually show an error do we? it seems like all we really do is indicate that the poster is no longer updating. is the error coming later?

is there another way to represent the poster URL which would indicate if it's reset or set? on error, instead of sending an error action, would it make sense to revert the URL to its previous state, or an empty string? if so, we could instead dispatch something like `setPosterUrl( previousURL )`

donnapep replied:
We do actually surface an error message to the user in the case where the poster could not be updated. To be clear, the new poster URL gets generated server-side and is sent back to the client on a successful update. If the poster updates without issue, the video editor modal closes and goes back to the Media Library modal, which will load the VideoPress player with the new poster showing.

In the event of an error occurring when updating a poster, we're not in a situation where the poster URL needs to be reverted because it was never changed in the first place.

As for the RESET_STATE action, that's there to clear out any existing state when opening the video editor modal again. Otherwise you might open it and see an error message if your first attempt at updating the poster wasn't successful.

dmsnell replied:
> We do actually surface an error message to the user

did I miss this? where does that happen?

dmsnell commented on client/state/data-layer/wpcom/videos/poster/index.js:
```diff
@@ -0,0 +1,69 @@
+};
+
+export const updatePosterUrl = ( { dispatch }, action, next, { poster } ) => {
+	dispatch( setPosterUrl( poster ) );
```
I guess `poster` is the name of the value coming from the API? it might help clarity here to rename in the arguments because it feels like `poster` ought to be an object here instead of a URL string…

```js
…, next, { poster as posterURL } ) => {
```

dmsnell commented on client/state/ui/editor/video-editor/reducer.js:
```diff
@@ -0,0 +1,88 @@
+export const uploadProgress = ( state = 0, action ) => {
+	switch ( action.type ) {
+		case VIDEO_EDITOR_SHOW_UPLOAD_PROGRESS:
+			return action.percentage;
```
some advice I'd share from the importer is that in some cases 100% is a value we never really see and it might be beneficial to fib a bit here. since 100% came at the very end, or never came because the `load` event occurred after something close like 96% I made it look like 100% names at an early value and this meant that people would see a positive confirmation that the processes was ending vs going from _processing_ to _oops I missed the end_.

this can also (maybe more appropriately) be handled in the UI

dmsnell commented:
@donnapep I'm still a bit leery about whether or not we need specific ERROR actions or even the RESET action. I think you said more error handling was on its way later, so please disregard if I'm overlooking that.

on the reset, what is a reset other than setting the URL to `''` or `null`? I think we can reuse our existing state if we think about the meaning of what we're storing. Often times I will go ahead and provide the action creator but reuse the existing action types. This helps keep our code bloat down and improves our reuse.

```js
const setUrl = url => ( { type: SET_URL, url } );

const resetUrl = () => setUrl( '' );

const url = ( state = null, { type, url } ) =>
	SET_URL === type
		? url
		: state;

const percentage = ( state = null, { type, percentage } ) => {
	if ( UPDATE === type ) {
	}
}
```

donnapep commented:
@dmsnell I've made some changes to the Redux actions and was able to eliminate the VIDEO_EDITOR_RESET_STATE action in https://github.com/Automattic/wp-calypso/pull/11633/commits/7e22d20b94b14db185e5ba77441a3efebee9cf05 as per your suggestion.

I did some other refactoring and, thanks to my amazing image annotation abilities, hereby present to you the following screen captures that hopefully help to better illustrate why each action is necessary and how it is used.

There are 5 actions added by the video editor, namely:

- VIDEO_EDITOR_CLOSE_MODAL (closes the video editor modal after a successful update - added by https://github.com/Automattic/wp-calypso/pull/11633/commits/f69706a57b582cbcaf64e04cbf9aee122d9e0023)
- VIDEO_EDITOR_SET_POSTER_URL (the poster URL is returned by the API and is used to update the thumbnail in the Media Library)
- VIDEO_EDITOR_SHOW_ERROR (shows an error if the poster could not be updated)
- VIDEO_EDITOR_SHOW_UPLOAD_PROGRESS (shows the progress bar while an image is being uploaded)
- VIDEO_EDITOR_UPDATE_POSTER (calls the API to update the poster and return the poster URL)

![upload-progress](https://cloud.githubusercontent.com/assets/1190420/24402685/a18661d4-1387-11e7-82a7-2ee815322634.jpg)

![show-error](https://cloud.githubusercontent.com/assets/1190420/24402691/a5a0fcb6-1387-11e7-85c8-5e69673a5132.jpg)

![set-poster-url](https://cloud.githubusercontent.com/assets/1190420/24402694/a7f92d94-1387-11e7-9d0b-7dad25d02dc4.jpg)

donnapep commented:
@dmsnell 🎉 Once again, thank you for your infinite patience and wisdom!
